pallets/pallet-bonded-coins/src/curves.rs

```diff
@@ -0,0 +1,253 @@
+use core::ops::{AddAssign, BitOrAssign, ShlAssign};
+use parity_scale_codec::{Decode, Encode, MaxEncodedLen};
+use scale_info::TypeInfo;
+use sp_arithmetic::ArithmeticError;
+use substrate_fixed::{
+	traits::{Fixed, FixedSigned, ToFixed},
+	transcendental::{exp, ln, sqrt},
+	types::I9F23,
+};
+
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo, MaxEncodedLen)]
+pub enum Curve<F> {
+	PolynomialFunction(PolynomialFunctionParameters<F>),
+	SquareRootBondingFunction(SquareRootFunctionParameters<F>),
+	LMSR(LMSRFunctionParameters<F>),
+}
+
+pub enum Operation<F> {
+	Mint(Vec<F>),
+	Burn(Vec<F>),
+}
+
+impl<F> Operation<F> {
+	pub fn inner_value(&self) -> &[F] {
+		match self {
+			Operation::Mint(x) => x,
+			Operation::Burn(x) => x,
+		}
+	}
+}
+
+impl<F> Curve<F>
+where
+	F: FixedSigned + PartialOrd<I9F23> + From<I9F23> + ToFixed,
+	<F as Fixed>::Bits: Copy + ToFixed + AddAssign + BitOrAssign + ShlAssign,
+{
+	fn calculate_accumulated_passive_issuance(passive_issuance: &[F]) -> F {
+		passive_issuance
+			.iter()
+			.fold(F::from_num(0), |sum, x| sum.saturating_add(*x))
+	}
+
+	fn calculate_low_high(op: Operation<F>, active_issuance_pre: F, active_issuance_post: F) -> (F, F) {
+		match op {
+			Operation::Burn(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+				)
+			}
+			Operation::Mint(passive) => {
+				let accumulated_passive_issuance = Self::calculate_accumulated_passive_issuance(&passive);
+				(
+					active_issuance_pre.saturating_add(accumulated_passive_issuance),
+					active_issuance_post.saturating_add(accumulated_passive_issuance),
+				)
+			}
+		}
+	}
+	pub fn calculate_cost(
+		&self,
+		active_issuance_pre: F,
+		active_issuance_post: F,
+		op: Operation<F>,
+	) -> Result<F, ArithmeticError> {
+		match self {
```

Comment:
Slippage is not handled here. The tx takes a max_costs/min_return parameter and if the collateral is not matching the tx will fail